### PR TITLE
[8.0] fix workflow loading

### PR DESCRIPTION
**Pull request by llacroix**
Prevent workflows from being used, if they depends on python modules that aren't yet loaded in sys.modules.

If the workflows are created without any reference to ir_model_data, they are loaded anyway as we do not have any idea if they depend on anything. Those are usually user created workflows... so the user should know.

https://github.com/odoo/odoo/pull/5703

### CODE REVIEW

**Comment by llacroix** [comment]
Problem related to this: 
https://github.com/OCA/knowledge/pull/30
https://github.com/OCA/knowledge/pull/25

**Comment by llacroix** [comment]
What's wrong with travis? Why is the file not present?

**Comment by hbrunn** [comment]
#129

**Comment by pedrobaeza** [comment]
Already merged, so you can make git rebase and then it has to be green.

File: openerp/workflow/service.py [hunk 1]
```diff
@@ -68,6 +69,15 @@ def delete(self):
...
 
     def create(self):
+
+        def check_module(module):
+            """Check if the module is loaded"""
+            return sys.modules.get('openerp.addons.%s' % module[0])
+
+        def to_record(module_record):
+            """Return the resource id"""
+            return (module_record[1], )
```
Comment (bwrsandman) [inline openerp/workflow/service.py]: Could you add `:param type module:`, `:param type module:` and `:returns:` elements to the doc. Namely, explain why you're fetching a certain index.

**Comment by bwrsandman** [comment]
@pedrobaeza no need for a rebase, I'll just restart the travis tests.

This can be done in Details of travis, the round arrow icon. No rebase needed :wink: 

capture d ecran de 2015-03-11 10 40 06 - 1

**Comment by llacroix** [comment]
It's green but I think runbot died before finishing. Runbot on odoo is green too. http://runbot.odoo.com/runbot/build/53432

**Comment by hbrunn** [comment]
:-1: this only fixes an issue with our travis script. It calls init on all modules, and runs tests only afterwards in a new process. This goes wrong in the case of workflows, but also in some other cases like modifying server actions and generally everywhere where installation modifies persistent data expecting the appropriate module is loaded.

I had the same problem in #129, which is why I pass `INSTALL_OPTIONS="--init base"` to prevent installing modules in https://github.com/OCA/maintainer-quality-tools/blob/master/travis/test_server#L167 ff.

**Comment by llacroix** [comment]
@hbrunn I wouldn't say that, I had the same issue locally on a simple upgrade. Workflows are going to get loaded even though the module that created it isn't yet loaded in memory. If the module that created the workflow adds new methods to the model and uses them in the module. It will fail to upgrade the database because the model might trigger a workflow that should not exist at the moment of loading.

I can make modules that reproduce the bug. But https://github.com/OCA/knowledge/pull/25 is a live example of the problem.

**Comment by hbrunn** [comment]
I understand the problem. But you'll only have it if there's an uninstallable module lurking around that has leftovers in the database or if you selectively run tests on some module(s) using workflows which are adapted by some other not yet loaded modules. Both cases should only happen to developers or other people hopefully knowing what they are doing

**Comment by llacroix** [comment]
@hbrunn So you're saying that it's fine if a workflow is getting triggered even though we are aware that its dependencies aren't yet loaded?

**Comment by hbrunn** [comment]
well, it's not fine of course. But I think it's the installation's administrator's job to take care that this doesn't happen. I prefer a crash telling that something is wrong a lot over silently ignoring those workflow steps, possibly creating bogus data

**Comment by llacroix** [comment]
@hbrunn well, I agree it would be better if we could delay a workflow trigger until the module is loaded. I'm not sure if we can do that though.

**Comment by llacroix** [comment]
http://runbot.odoo.com/runbot/build/56926 tests are passing here (runbot)

**Comment by llacroix** [comment]
@hbrunn How about that? now it loads the missing modules if they aren't yet loaded.. if travis pass tests, I'll cleanup what isn't used anymore.

**Comment by pedrobaeza** [comment]
Closing this as upstream is was rejected due to several reasons that you can read in the other thread.

**Pull request closed without merging**